File: app/src/main/java/com/yvrun/officeprocess/mvp/view/BaseMvpFragment.java
package com.yvrun.officeprocess.mvp.view;

import androidx.lifecycle.Lifecycle;

import com.uber.autodispose.AutoDispose;
import com.uber.autodispose.AutoDisposeConverter;
import com.uber.autodispose.android.lifecycle.AndroidLifecycleScopeProvider;
import com.yvrun.officeprocess.mvp.model.BaseModel;
import com.yvrun.officeprocess.mvp.presenter.BasePresenter;

import javax.inject.Inject;

import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.HasAndroidInjector;
import dagger.android.support.AndroidSupportInjection;

public abstract class BaseMvpFragment<P extends BasePresenter,M extends BaseModel> extends BaseFragment implements IBaseView, HasAndroidInjector {

    @Inject
    protected P mPresenter;
    @Inject
    DispatchingAndroidInjector<Object> androidInjector;

    @Override
    protected void bindMvp() {
        AndroidSupportInjection.inject(this);
    }

    @Override
    public AndroidInjector<Object> androidInjector() {
        return androidInjector;
    }

    /**
     * 绑定生命周期 防止MVP内存泄漏
     *
     * @param <T> 数据类型
     * @return 数据流
     */
    @Override
    public <T> AutoDisposeConverter<T> bindAutoDispose() {
        return AutoDispose.autoDisposable(AndroidLifecycleScopeProvider
                .from(this,Lifecycle.Event.ON_DESTROY));
    }
}

File: kotlinappother/src/main/java/com/chengzhen/kotlinappother/mvp/view/BaseMvpFragment.kt
package com.chengzhen.kotlinappother.mvp.view

import androidx.lifecycle.Lifecycle

import com.chengzhen.kotlinappother.base.BaseFragment
import com.chengzhen.kotlinappother.mvp.presenter.BasePresenter
import com.uber.autodispose.AutoDispose
import com.uber.autodispose.AutoDisposeConverter
import com.uber.autodispose.android.lifecycle.AndroidLifecycleScopeProvider

import javax.inject.Inject

import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasAndroidInjector
import dagger.android.support.AndroidSupportInjection

abstract class BaseMvpFragment<P : BasePresenter<*, *>> : BaseFragment(), IBaseView, HasAndroidInjector {

    @Inject
    var mPresenter: P? = null
    @Inject
    internal var androidInjector: DispatchingAndroidInjector<Any>? = null

    override fun bindMvp() {
        AndroidSupportInjection.inject(this)
    }

    override fun androidInjector(): AndroidInjector<Any>? {
        return androidInjector
    }

    /**
     * 绑定生命周期 防止MVP内存泄漏
     *
     * @param <T> 数据类型
     * @return 数据流
    </T> */
    override fun <T> bindAutoDispose(): AutoDisposeConverter<T> {
        return AutoDispose.autoDisposable(AndroidLifecycleScopeProvider
                .from(this, Lifecycle.Event.ON_DESTROY))
    }
}
